### PR TITLE
ENG-12557: add CRUD operations for AWS IAM AuthN integration

### PR DESCRIPTION
**Pull request by ricardorey10**
## Description of the change

Recently, [an API](https://github.com/cyralinc/integrations/blob/main/proto/aws-iam/aws-iam.proto#L16) for CRUD'ing AWS IAM integrations was added. This PR adds the necessary Terraform code to support that.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing
Tested locally in a k8s CP. I'll add some unit tests too.

### CODE REVIEW

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
+package cyral
```
Comment (ricardorey10) [inline cyral/resource_cyral_integration_aws_iam_authn.go:1]: @VictorGFM @Yowgf I still have to add unit tests, but I wanted some early feedback before doing that :grin:

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+			d.Id(),
+		)
+	},
+	NewResponseData: func(_ *schema.ResourceData) ResponseData {
+		return &APIWrapper{}
+	},
+	RequestErrorHandler: &ReadIgnoreHttpNotFound{resName: "AWS IAM AuthN Integration"},
+}
+
+func resourceIntegrationAWSIAMAuthN() *schema.Resource {
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:83]:
```suggestion
func resourceIntegrationAWSIAM() *schema.Resource {
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+		return fmt.Sprintf(
+			"https://%s/v1/integrations/aws/iam/%s",
+			c.ControlPlane,
+			d.Id(),
+		)
+	},
+	NewResponseData: func(_ *schema.ResourceData) ResponseData {
+		return &APIWrapper{}
+	},
+	RequestErrorHandler: &ReadIgnoreHttpNotFound{resName: "AWS IAM AuthN Integration"},
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:80]:
```suggestion
	RequestErrorHandler: &ReadIgnoreHttpNotFound{resName: "AWS IAM Integration"},
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+				Name:       "AWSIAMIntegrationCreate",
+				HttpMethod: http.MethodPost,
+				CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+					return fmt.Sprintf("https://%s/v1/integrations/aws/iam", c.ControlPlane)
+				},
+				NewResourceData: func() ResourceData {
+					return &APIWrapper{}
+				},
+				NewResponseData: func(_ *schema.ResourceData) ResponseData {
+					return &CreateAWSIAMIntegrationResponse{}
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:97]: Lets use `IDBasedResponse` instead:
```suggestion
					return &IDBasedResponse{}
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+type CreateAWSIAMIntegrationResponse struct {
+	ID string `json:"id"`
+}
+
+func (c *CreateAWSIAMIntegrationResponse) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(c.ID)
+	return nil
+}
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:65]:
```suggestion
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type APIWrapper struct {
+	Integration *AWSIAMIntegrationResource `json:"iamIntegration"`
+}
+
+type AWSIAMIntegrationResource struct {
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:16]:
```suggestion
type AWSIAMIntegration struct {
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type APIWrapper struct {
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:12]:
```suggestion
type AWSIAMIntegrationWrapper struct {
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+		},
+
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Description: "Terraform ID of this resource",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+
+			"name": {
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:156]: Lets create constants for the schema keys (see `resource_cyral_repository` for an example) and use them on d.Set and d.Get and other places that might be needed.
```suggestion
			AWSIAMIntegrationNameKey: {
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+		return fmt.Errorf("error setting 'name': %w", err)
+	}
+
+	if err := d.Set("description", integration.Description); err != nil {
+		return fmt.Errorf("error setting 'description': %w", err)
+	}
+
+	if err := d.Set("arns", integration.IAMRoleARNs); err != nil {
+		return fmt.Errorf("error setting 'arns': %w", err)
+	}
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:38]:
```suggestion
	if err := d.Set(RoleARNsKey, integration.IAMRoleARNs); err != nil {
		return fmt.Errorf("error setting '%s': %w", err, RoleARNsKey)
	}
```

File: cyral/resource_cyral_integration_aws_iam_authn_test.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+		Steps: []resource.TestStep{
+			testCreate,
+			testUpdate1,
+			testUpdate2,
+			testUpdate3,
+			testUpdate4,
+			{
+				ImportState:       true,
+				ImportStateVerify: true,
+				ResourceName:      "cyral_integration_aws_iam_authn.main_test",
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn_test.go:33]:
```suggestion
				ResourceName:      "cyral_integration_aws_iam.main_test",
```

File: cyral/resource_cyral_integration_aws_iam_authn_test.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+const (
+	arn0 = "arn:aws:iam::421552459763:role/Role1"
+	arn1 = "arn:aws:iam::421552459763:role/Role2"
+)
+
+func TestIntegrationAWSIAMAuthN(t *testing.T) {
+	testCreate := getCreateStep("main_test")
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn_test.go:16]: Lets use the `accTestName` function to make sure resources generated by terraform tests can be identified.
```suggestion
        resourceName := accTestName("integration-aws-iam", "main_test")
	testCreate := getCreateStep(resourceName)
```

File: docs/data-sources/datalabel.md [hunk 12]
```diff
@@ -1,6 +1,6 @@
 ---
 # generated by https://github.com/hashicorp/terraform-plugin-docs
-page_title: "cyral_datalabel Data Source - cyral"
+page_title: "cyral_datalabel Data Source - terraform-provider-cyral"
```
Comment (VictorGFM) [inline docs/data-sources/datalabel.md:3]: Lets rebase the PR on top of main to solve these doc diffs.

**Review by VictorGFM — COMMENTED** [review]
Hey @ricardorey10, changes are looking good. Here are a few comments for your consideration:

File: cyral/resource_cyral_integration_aws_iam_authn_test.go (hunk 11 above)
Comment (VictorGFM) [reply cyral/resource_cyral_integration_aws_iam_authn_test.go]: @ricardorey10 I just noticed that my suggestion above does not make sense. We don't need to use the accTestName for the resourceName if that's only used to identify the terraform resource name. Sorry for that, please feel free to ignore this suggestion.

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 13]
```diff
@@ -0,0 +1,164 @@
+package cyral
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:1]: Lets also rename the file name to `resource_cyral_integration_aws_iam.go` instead (same for the test file).

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+	},
+	NewResponseData: func(_ *schema.ResourceData) ResponseData {
+		return &AWSIAMIntegrationWrapper{}
+	},
+	RequestErrorHandler: &ReadIgnoreHttpNotFound{resName: "AWS IAM Integration"},
+}
+
+func resourceIntegrationAWSIAM() *schema.Resource {
+	return &schema.Resource{
+		Description: "Authenticate users based on AWS IAM credentials",
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:81]:
```suggestion
		Description: "Authenticate users based on AWS IAM credentials.",
```

File: cyral/resource_cyral_integration_aws_iam_authn.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+				},
+			},
+		),
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"id": {
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_aws_iam_authn.go:135]:
```suggestion
			IDKey: {
```

**Review by VictorGFM — APPROVED** [review]
Thanks for working on this @ricardorey10, Im already approving the PR but please make sure to address the comments below before merging.

**Pull request merged**